File: parent-promotion-caps-override-coupon/python/find_capped_promotion_codes.py
"""Find BigCommerce coupon codes silently gated by their parent promotion's cap.

In the Promotions v3 model, a coupon code is a child resource nested under a
parent Promotion (/v3/promotions/{promotionId}/codes/{codeId}), and both levels
carry independent max_uses/current_uses counters. BigCommerce enforces both at
checkout, and the promotion-level cap is the outer gate: even if a code's own
max_uses has plenty of headroom, the shopper gets "invalid coupon code" once the
parent promotion's aggregate current_uses reaches its max_uses. This job lists
every ENABLED promotion, pages its coupon codes, and flags any code where the
promotion is already exhausted or where the code's own remaining uses exceed
what the promotion has left. It never writes to a promotion's cap by default.
Raising a merchant's deliberate cap is a business decision, so a write only
happens behind an explicit --apply flag with DRY_RUN=false, and it always
prints the proposed diff first.

Guide: https://www.allanninal.dev/bigcommerce/parent-promotion-caps-override-coupon/
"""
import os
import sys
import logging

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("find_capped_promotion_codes")

STORE_HASH = os.environ["BIGCOMMERCE_STORE_HASH"]
ACCESS_TOKEN = os.environ["BIGCOMMERCE_ACCESS_TOKEN"]
API_BASE = f"https://api.bigcommerce.com/stores/{STORE_HASH}/v3"
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

HEADERS = {
    "X-Auth-Token": ACCESS_TOKEN,
    "Content-Type": "application/json",
    "Accept": "application/json",
}


def bc_get_page(path, params=None):
    r = requests.get(f"{API_BASE}{path}", headers=HEADERS, params=params or {}, timeout=30)
    r.raise_for_status()
    body = r.json()
    return body.get("data", []), body.get("meta", {}).get("pagination", {})


def bc_get_all(path, params=None):
    page = 1
    items = []
    while True:
        data, pagination = bc_get_page(path, {**(params or {}), "page": page, "limit": 250})
        items.extend(data)
        total_pages = pagination.get("total_pages", 1)
        if page >= total_pages:
            return items
        page += 1


def bc_put(path, body):
    r = requests.put(f"{API_BASE}{path}", headers=HEADERS, json=body, timeout=30)
    r.raise_for_status()
    return r.json()


def find_capped_out_codes(promotion: dict, codes: list) -> list:
    """Pure decision. No network, no side effects.

    promotion={id,max_uses,current_uses,status}
    codes=[{id,code,max_uses,current_uses}]

    promo_remaining = None if promotion["max_uses"] == 0 else
        max(promotion["max_uses"] - promotion["current_uses"], 0)
    code_remaining = None if code["max_uses"] == 0 else
        max(code["max_uses"] - code["current_uses"], 0)

    reason is "promotion_exhausted" if promo_remaining == 0,
    "promotion_cap_lower_than_code" if promo_remaining is not None and
        (code_remaining is None or code_remaining > promo_remaining),
    otherwise "ok".
    """
    promo_max = promotion.get("max_uses", 0) or 0
    promo_current = promotion.get("current_uses", 0) or 0
    promo_remaining = None if promo_max == 0 else max(promo_max - promo_current, 0)

    results = []
    for code in codes:
        code_max = code.get("max_uses", 0) or 0
        code_current = code.get("current_uses", 0) or 0
        code_remaining = None if code_max == 0 else max(code_max - code_current, 0)

        if promo_remaining == 0:
            reason = "promotion_exhausted"
        elif promo_remaining is not None and (
            code_remaining is None or code_remaining > promo_remaining
        ):
            reason = "promotion_cap_lower_than_code"
        else:
            reason = "ok"

        results.append({
            "code_id": code.get("id"),
            "code": code.get("code"),
            "reason": reason,
            "promotion_remaining": promo_remaining,
            "code_remaining": code_remaining,
        })
    return results


def enabled_coupon_promotions():
    promotions = bc_get_all("/promotions", {"status": "ENABLED"})
    return [p for p in promotions if "COUPON" in (p.get("redemption_type") or "")]


def promotion_codes(promotion_id):
    return bc_get_all(f"/promotions/{promotion_id}/codes")


def propose_raised_cap(promotion, target_max_uses):
    return {
        "promotion_id": promotion["id"],
        "from_max_uses": promotion.get("max_uses", 0),
        "to_max_uses": target_max_uses,
    }


def apply_raised_cap(promotion_id, target_max_uses):
    return bc_put(f"/promotions/{promotion_id}", {"max_uses": target_max_uses})


def run(apply_fix=False):
    flagged = 0
    checked = 0

    for promotion in enabled_coupon_promotions():
        codes = promotion_codes(promotion["id"])
        annotated = find_capped_out_codes(promotion, codes)
        checked += len(annotated)

        problem_codes = [c for c in annotated if c["reason"] != "ok"]
        if not problem_codes:
            continue

        max_code_max_uses = max((c.get("max_uses", 0) or 0) for c in codes) if codes else 0
        target_max_uses = max(max_code_max_uses, promotion.get("max_uses", 0) or 0)

        for c in problem_codes:
            log.warning(
                "promotion_id=%s promotion_name=%s code_id=%s code=%s reason=%s "
                "promotion_remaining=%s code_remaining=%s",
                promotion["id"], promotion.get("name"), c["code_id"], c["code"],
                c["reason"], c["promotion_remaining"], c["code_remaining"],
            )
            flagged += 1

        if apply_fix and target_max_uses != (promotion.get("max_uses", 0) or 0):
            diff = propose_raised_cap(promotion, target_max_uses)
            log.info("Proposed fix: %s (%s)", diff, "dry run" if DRY_RUN else "applying")
            if not DRY_RUN:
                apply_raised_cap(promotion["id"], target_max_uses)

    log.info("Done. %d code(s) checked, %d code(s) flagged.", checked, flagged)


if __name__ == "__main__":
    run(apply_fix="--apply" in sys.argv)

File: parent-promotion-caps-override-coupon/node/find-capped-promotion-codes.js
/**
 * Find BigCommerce coupon codes silently gated by their parent promotion's cap.
 *
 * In the Promotions v3 model, a coupon code is a child resource nested under a
 * parent Promotion (/v3/promotions/{promotionId}/codes/{codeId}), and both levels
 * carry independent max_uses/current_uses counters. BigCommerce enforces both at
 * checkout, and the promotion-level cap is the outer gate: even if a code's own
 * max_uses has plenty of headroom, the shopper gets "invalid coupon code" once the
 * parent promotion's aggregate current_uses reaches its max_uses. This job lists
 * every ENABLED promotion, pages its coupon codes, and flags any code where the
 * promotion is already exhausted or where the code's own remaining uses exceed
 * what the promotion has left. It never writes to a promotion's cap by default.
 * Raising a merchant's deliberate cap is a business decision, so a write only
 * happens behind an explicit --apply flag with DRY_RUN=false, and it always
 * prints the proposed diff first.
 *
 * Guide: https://www.allanninal.dev/bigcommerce/parent-promotion-caps-override-coupon/
 */
import { pathToFileURL } from "node:url";

const STORE_HASH = process.env.BIGCOMMERCE_STORE_HASH || "example_hash";
const ACCESS_TOKEN = process.env.BIGCOMMERCE_ACCESS_TOKEN || "bc_dummy";
const API_BASE = `https://api.bigcommerce.com/stores/${STORE_HASH}/v3`;
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const HEADERS = {
  "X-Auth-Token": ACCESS_TOKEN,
  "Content-Type": "application/json",
  Accept: "application/json",
};

/**
 * Pure decision. No network, no side effects.
 *
 * promotion={id,max_uses,current_uses,status}
 * codes=[{id,code,max_uses,current_uses}]
 *
 * promoRemaining = null if promotion.max_uses === 0 else
 *   max(promotion.max_uses - promotion.current_uses, 0)
 * codeRemaining = null if code.max_uses === 0 else
 *   max(code.max_uses - code.current_uses, 0)
 *
 * reason is "promotion_exhausted" if promoRemaining === 0,
 * "promotion_cap_lower_than_code" if promoRemaining !== null and
 *   (codeRemaining === null or codeRemaining > promoRemaining),
 * otherwise "ok".
 */
export function findCappedOutCodes(promotion, codes) {
  const promoMax = promotion.max_uses || 0;
  const promoCurrent = promotion.current_uses || 0;
  const promoRemaining = promoMax === 0 ? null : Math.max(promoMax - promoCurrent, 0);

  return codes.map((code) => {
    const codeMax = code.max_uses || 0;
    const codeCurrent = code.current_uses || 0;
    const codeRemaining = codeMax === 0 ? null : Math.max(codeMax - codeCurrent, 0);

    let reason;
    if (promoRemaining === 0) {
      reason = "promotion_exhausted";
    } else if (
      promoRemaining !== null &&
      (codeRemaining === null || codeRemaining > promoRemaining)
    ) {
      reason = "promotion_cap_lower_than_code";
    } else {
      reason = "ok";
    }

    return {
      code_id: code.id,
      code: code.code,
      reason,
      promotion_remaining: promoRemaining,
      code_remaining: codeRemaining,
    };
  });
}

async function bcGetPage(path, params = {}) {
  const url = new URL(`${API_BASE}${path}`);
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== null) url.searchParams.set(key, value);
  }
  const res = await fetch(url, { headers: HEADERS });
  if (!res.ok) throw new Error(`BigCommerce ${res.status}`);
  const body = await res.json();
  return [body.data || [], (body.meta || {}).pagination || {}];
}

async function bcGetAll(path, params = {}) {
  let page = 1;
  const items = [];
  while (true) {
    const [data, pagination] = await bcGetPage(path, { ...params, page, limit: 250 });
    items.push(...data);
    const totalPages = pagination.total_pages || 1;
    if (page >= totalPages) return items;
    page += 1;
  }
}

async function bcPut(path, body) {
  const res = await fetch(`${API_BASE}${path}`, {
    method: "PUT",
    headers: HEADERS,
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`BigCommerce ${res.status}`);
  return res.json();
}

async function enabledCouponPromotions() {
  const promotions = await bcGetAll("/promotions", { status: "ENABLED" });
  return promotions.filter((p) => (p.redemption_type || "").includes("COUPON"));
}

async function promotionCodes(promotionId) {
  return bcGetAll(`/promotions/${promotionId}/codes`);
}

function proposeRaisedCap(promotion, targetMaxUses) {
  return {
    promotion_id: promotion.id,
    from_max_uses: promotion.max_uses || 0,
    to_max_uses: targetMaxUses,
  };
}

async function applyRaisedCap(promotionId, targetMaxUses) {
  return bcPut(`/promotions/${promotionId}`, { max_uses: targetMaxUses });
}

export async function run(applyFix = false) {
  let flagged = 0;
  let checked = 0;

  for (const promotion of await enabledCouponPromotions()) {
    const codes = await promotionCodes(promotion.id);
    const annotated = findCappedOutCodes(promotion, codes);
    checked += annotated.length;

    const problemCodes = annotated.filter((c) => c.reason !== "ok");
    if (!problemCodes.length) continue;

    const maxCodeMaxUses = codes.length ? Math.max(...codes.map((c) => c.max_uses || 0)) : 0;
    const targetMaxUses = Math.max(maxCodeMaxUses, promotion.max_uses || 0);

    for (const c of problemCodes) {
      console.warn(
        `promotion_id=${promotion.id} promotion_name=${promotion.name} code_id=${c.code_id} ` +
        `code=${c.code} reason=${c.reason} promotion_remaining=${c.promotion_remaining} ` +
        `code_remaining=${c.code_remaining}`
      );
      flagged += 1;
    }

    if (applyFix && targetMaxUses !== (promotion.max_uses || 0)) {
      const diff = proposeRaisedCap(promotion, targetMaxUses);
      console.log(`Proposed fix: ${JSON.stringify(diff)} (${DRY_RUN ? "dry run" : "applying"})`);
      if (!DRY_RUN) await applyRaisedCap(promotion.id, targetMaxUses);
    }
  }

  console.log(`Done. ${checked} code(s) checked, ${flagged} code(s) flagged.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  const applyFix = process.argv.includes("--apply");
  run(applyFix).catch((err) => { console.error(err); process.exit(1); });
}
